Convert README to Markdown (#125)
python/cherry-picker: README.md

default_branch  Project's default branch name,
                e.g "devel" for https://github.com/ansible/ansible
                ("main" by default)
```

To customize the tool for used by other project:

1. Create a file called `.cherry_picker.toml` in the project's root
   folder (alongside with `.git` folder).

2. Add `team`, `repo`, `fix_commit_msg`, `check_sha` and
   `default_branch` config values as described above.

3. Use `git add .cherry_picker.toml` / `git commit` to add the config
   into Git.

4. Add `cherry_picker` to development dependencies or install it
   by `pip install cherry_picker`

5. Now everything is ready, use `cherry_picker <commit_sha> <branch1>
   <branch2>` for cherry-picking changes from `<commit_sha>` into
   maintenance branches.
   Branch name should contain at least major and minor version numbers
   and may have some prefix or suffix.
   Only the first version-like substring is matched when the version
   is extracted from branch name.

### Demo

- Installation: https://asciinema.org/a/125254

- Backport: https://asciinema.org/a/125256

### Example

For example, to cherry-pick `6de2b7817f-some-commit-sha1-d064` into

> `3.5` and `3.6`, run the following command from the cloned CPython

Reviewer: ```suggestion
`3.12` and `3.11`, run the following command from the cloned CPython
```

directory:

```console

> (venv) $ cherry_picker 6de2b7817f-some-commit-sha1-d064 3.5 3.6

Reviewer: ```suggestion
(venv) $ cherry_picker 6de2b7817f-some-commit-sha1-d064 3.12 3.11
```

```

What this will do:

```console
(venv) $ git fetch upstream

> (venv) $ git checkout -b backport-6de2b78-3.5 upstream/3.5
> (venv) $ git cherry-pick -x 6de2b7817f-some-commit-sha1-d064
> (venv) $ git push origin backport-6de2b78-3.5
> (venv) $ git checkout main
> (venv) $ git branch -D backport-6de2b78-3.5

Reviewer: ```suggestion
(venv) $ git checkout -b backport-6de2b78-3.12 upstream/3.12
(venv) $ git cherry-pick -x 6de2b7817f-some-commit-sha1-d064
(venv) $ git push origin backport-6de2b78-3.12
(venv) $ git checkout main
(venv) $ git branch -D backport-6de2b78-3.12
```

> (venv) $ git checkout -b backport-6de2b78-3.6 upstream/3.6
> (venv) $ git cherry-pick -x 6de2b7817f-some-commit-sha1-d064
> (venv) $ git push origin backport-6de2b78-3.6
> (venv) $ git checkout main
> (venv) $ git branch -D backport-6de2b78-3.6

Reviewer: ```suggestion
(venv) $ git checkout -b backport-6de2b78-3.11 upstream/3.11
(venv) $ git cherry-pick -x 6de2b7817f-some-commit-sha1-d064
(venv) $ git push origin backport-6de2b78-3.11
(venv) $ git checkout main
(venv) $ git branch -D backport-6de2b78-3.11
```

```

In case of merge conflicts or errors, the following message will be displayed:

```
Failed to cherry-pick 554626ada769abf82a5dabe6966afa4265acb6a6 into 2.7 :frowning_face:
... Stopping here.

To continue and resolve the conflict:
    $ cherry_picker --status  # to find out which files need attention
    # Fix the conflict
    $ cherry_picker --status  # should now say 'all conflict fixed'
    $ cherry_picker --continue

To abort the cherry-pick and cleanup:
    $ cherry_picker --abort
```

Passing the `--dry-run` option will cause the script to print out all the
steps it would execute without actually executing any of them. For example:

```console

> $ cherry_picker --dry-run --pr-remote pr 1e32a1be4a1705e34011770026cb64ada2d340b5 3.6 3.5
> Dry run requested, listing expected command sequence
> fetching upstream ...
> dry_run: git fetch origin
> Now backporting '1e32a1be4a1705e34011770026cb64ada2d340b5' into '3.6'
> dry_run: git checkout -b backport-1e32a1b-3.6 origin/3.6
> dry_run: git cherry-pick -x 1e32a1be4a1705e34011770026cb64ada2d340b5
> dry_run: git push pr backport-1e32a1b-3.6
> dry_run: Create new PR: https://github.com/python/cpython/compare/3.6...ncoghlan:backport-1e32a1b-3.6?expand=1
> dry_run: git checkout main
> dry_run: git branch -D backport-1e32a1b-3.6
> Now backporting '1e32a1be4a1705e34011770026cb64ada2d340b5' into '3.5'
> dry_run: git checkout -b backport-1e32a1b-3.5 origin/3.5
> dry_run: git cherry-pick -x 1e32a1be4a1705e34011770026cb64ada2d340b5
> dry_run: git push pr backport-1e32a1b-3.5
> dry_run: Create new PR: https://github.com/python/cpython/compare/3.5...ncoghlan:backport-1e32a1b-3.5?expand=1
> dry_run: git checkout main
> dry_run: git branch -D backport-1e32a1b-3.5

Reviewer: ```suggestion
$ cherry_picker --dry-run --pr-remote pr 1e32a1be4a1705e34011770026cb64ada2d340b5 3.12 3.11
Dry run requested, listing expected command sequence
fetching upstream ...
dry_run: git fetch origin
Now backporting '1e32a1be4a1705e34011770026cb64ada2d340b5' into '3.12'
dry_run: git checkout -b backport-1e32a1b-3.12 origin/3.12
dry_run: git cherry-pick -x 1e32a1be4a1705e34011770026cb64ada2d340b5
dry_run: git push pr backport-1e32a1b-3.12
dry_run: Create new PR: https://github.com/python/cpython/compare/3.12...ncoghlan:backport-1e32a1b-3.12?expand=1
dry_run: git checkout main
dry_run: git branch -D backport-1e32a1b-3.12
Now backporting '1e32a1be4a1705e34011770026cb64ada2d340b5' into '3.11'
dry_run: git checkout -b backport-1e32a1b-3.11 origin/3.11
dry_run: git cherry-pick -x 1e32a1be4a1705e34011770026cb64ada2d340b5
dry_run: git push pr backport-1e32a1b-3.11
dry_run: Create new PR: https://github.com/python/cpython/compare/3.11...ncoghlan:backport-1e32a1b-3.11?expand=1
dry_run: git checkout main
dry_run: git branch -D backport-1e32a1b-3.11
```

This will generate pull requests through a remote other than `origin`
(e.g. `pr`)

### `--upstream-remote` option

This will generate branches from a remote other than `upstream`/`origin`
(e.g. `python`)

### `--status` option

This will do `git status` for the CPython directory.

### `--abort` option

Cancels the current cherry-pick and cleans up the cherry-pick branch.

### `--continue` option

Continues the current cherry-pick, commits, pushes the current branch to
`origin`, opens the PR page, and cleans up the branch.

### `--no-push` option

Changes won't be pushed to remote.  This allows you to test and make additional
changes.  Once you're satisfied with local changes, use `--continue` to complete
the backport, or `--abort` to cancel and clean up the branch.  You can also
cherry-pick additional commits, by:

```console
$ git cherry-pick -x <commit_sha1>
```

### `--no-auto-pr` option

PR creation page won't be automatically opened in the web browser or
if GH_AUTH is set, the PR won't be automatically opened through API.
This can be useful if your terminal is not capable of opening a useful web browser,
or if you use cherry-picker with a different Git hosting than GitHub.

> ### `--config-path` option

Reviewer: This might look better as a list/table, rather than a series of titles+paragraph.

Author: Something like this?

In another branch: https://github.com/hugovk/cherry-picker/tree/convert-readme-to-md2

But compare with: 

https://github.com/hugovk/cherry-picker/tree/convert-readme-to-md?tab=readme-ov-file#--no-push-option

I used `<br><br>` to replicate a new paragraph, and we can only use single backticks around the command instead of three, so it's no longer copyable.

Allows to override default config file path
(`<PROJ-ROOT>/.cherry_picker.toml`) with a custom one. This allows cherry_picker
to backport projects other than CPython.

## Creating pull requests

When a cherry-pick was applied successfully, this script will open up a browser
tab that points to the pull request creation page.

The url of the pull request page looks similar to the following:

```

> https://github.com/python/cpython/compare/3.5...<username>:backport-6de2b78-3.5?expand=1

Reviewer: ```suggestion
https://github.com/python/cpython/compare/3.12...<username>:backport-6de2b78-3.12?expand=1
```
